kafka sink not robust to admin tasks

Kafka admin task meet error, cause the whole changefeed restart. 

```
[2023/02/09 15:42:53.245 +00:00] [ERROR] [manager.go:199] ["Worker handles sink task failed"] [namespace=default] [changefeed=kafka-scale-big-table] [error="write tcp 10.233.84.234:54958->10.233.122.44:9092: write: broken pipe"] [errorVerbose="write tcp 10.233.84.234:54958->10.233.122.44:9092: write: broken pipe\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/cdc/sink/mq/manager.(*kafkaTopicManager).GetPartitionNum\n\tgithub.com/pingcap/tiflow/cdc/sink/mq/manager/kafka_manager.go:70\ngithub.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq.(*dmlSink).WriteEvents\n\tgithub.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq/mq_dml_sink.go:120\ngithub.com/pingcap/tiflow/cdc/sinkv2/tablesink.(*EventTableSink[...]).UpdateResolvedTs\n\tgithub.com/pingcap/tiflow/cdc/sinkv2/tablesink/table_sink_impl.go:115\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*tableSinkWrapper).updateResolvedTs\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_wrapper.go:170\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).advanceTableSink\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:459\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask.func2\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:167\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask.func3\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:214\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:320\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTasks\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:76\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*SinkManager).startWorkers.func1\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/manager.go:197\nruntime.goexit\n\truntime/asm_amd64.s:1594"]
[2023/02/09 15:42:53.252 +00:00] [ERROR] [processor.go:745] ["error on running processor"] [capture=47d9ab45-1bae-4061-96fd-d0f2763e411d] [namespace=default] [changefeed=kafka-scale-big-table] [error="write tcp 10.233.84.234:54958->10.233.122.44:9092: write: broken pipe"] [errorVerbose="write tcp 10.233.84.234:54958->10.233.122.44:9092: write: broken pipe\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/cdc/sink/mq/manager.(*kafkaTopicManager).GetPartitionNum\n\tgithub.com/pingcap/tiflow/cdc/sink/mq/manager/kafka_manager.go:70\ngithub.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq.(*dmlSink).WriteEvents\n\tgithub.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq/mq_dml_sink.go:120\ngithub.com/pingcap/tiflow/cdc/sinkv2/tablesink.(*EventTableSink[...]).UpdateResolvedTs\n\tgithub.com/pingcap/tiflow/cdc/sinkv2/tablesink/table_sink_impl.go:115\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*tableSinkWrapper).updateResolvedTs\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_wrapper.go:170\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).advanceTableSink\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:459\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask.func2\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:167\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask.func3\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:214\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:320\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTasks\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:76\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*SinkManager).startWorkers.func1\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/manager.go:197\nruntime.goexit\n\truntime/asm_amd64.s:1594"]
[2023/02/09 15:42:53.252 +00:00] [ERROR] [processor.go:539] ["run processor failed"] [capture=47d9ab45-1bae-4061-96fd-d0f2763e411d] [namespace=default] [changefeed=kafka-scale-big-table] [error="write tcp 10.233.84.234:54958->10.233.122.44:9092: write: broken pipe"] [errorVerbose="write tcp 10.233.84.234:54958->10.233.122.44:9092: write: broken pipe\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/tiflow/cdc/sink/mq/manager.(*kafkaTopicManager).GetPartitionNum\n\tgithub.com/pingcap/tiflow/cdc/sink/mq/manager/kafka_manager.go:70\ngithub.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq.(*dmlSink).WriteEvents\n\tgithub.com/pingcap/tiflow/cdc/sinkv2/eventsink/mq/mq_dml_sink.go:120\ngithub.com/pingcap/tiflow/cdc/sinkv2/tablesink.(*EventTableSink[...]).UpdateResolvedTs\n\tgithub.com/pingcap/tiflow/cdc/sinkv2/tablesink/table_sink_impl.go:115\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*tableSinkWrapper).updateResolvedTs\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_wrapper.go:170\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).advanceTableSink\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:459\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask.func2\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:167\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask.func3\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:214\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTask\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:320\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*sinkWorker).handleTasks\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/table_sink_worker.go:76\ngithub.com/pingcap/tiflow/cdc/processor/sinkmanager.(*SinkManager).startWorkers.func1\n\tgithub.com/pingcap/tiflow/cdc/processor/sinkmanager/manager.go:197\nruntime.goexit\n\truntime/asm_amd64.s:1594"]
```

--- a/pkg/sink/kafka/admin.go
+++ b/pkg/sink/kafka/admin.go
@@ -16,6 +16,8 @@ package kafka
 import (
 	"context"
 	"errors"
+	"io"
+	"net"
 	"strconv"
 	"sync"
 	"syscall"
@@ -80,6 +82,12 @@ func (a *saramaAdminClient) queryClusterWithRetry(ctx context.Context, query fun
 		}
 
 		if !errors.Is(err, syscall.EPIPE) {
+			return err
+		}
+		if !errors.Is(err, net.ErrClosed) {
+			return err
+		}
+		if !errors.Is(err, io.EOF) {
 			return err
 		}
 

--- a/pkg/sink/kafka/admin.go
+++ b/pkg/sink/kafka/admin.go
@@ -17,7 +17,6 @@ import (
 	"context"
 	"errors"
 	"strconv"
-	"strings"
 	"sync"
 	"syscall"
 
@@ -119,21 +118,12 @@ func (a *saramaAdminClient) GetCoordinator(ctx context.Context) (int, error) {
 		controllerID int32
 		err          error
 	)
-	err = retry.Do(ctx, func() error {
-		a.mu.Lock()
-		defer a.mu.Unlock()
+
+	query := func() error {
 		_, controllerID, err = a.client.DescribeCluster()
-		if err == nil {
-			return nil
-		}
-
-		if !errors.Is(err, syscall.EPIPE) {
-			return err
-		}
-
-		return a.reset()
-	}, retry.WithBackoffBaseDelay(defaultRetryBackoff), retry.WithMaxTries(defaultRetryMaxTries))
-
+		return err
+	}
+	err = a.queryClusterWithRetry(ctx, query)
 	return int(controllerID), err
 }
 
@@ -147,23 +137,15 @@ func (a *saramaAdminClient) GetBrokerConfig(
 	}
 
 	var configEntries []sarama.ConfigEntry
-	err = retry.Do(ctx, func() error {
-		a.mu.Lock()
-		defer a.mu.Unlock()
+	query := func() error {
 		configEntries, err = a.client.DescribeConfig(sarama.ConfigResource{
 			Type:        sarama.BrokerResource,
 			Name:        strconv.Itoa(controller),
 			ConfigNames: []string{configName},
 		})
-		if err == nil {
-			return nil
-		}
-
-		if !errors.Is(err, syscall.EPIPE) {
-			return err
-		}
-		return a.reset()
-	}, retry.WithBackoffBaseDelay(defaultRetryBackoff), retry.WithMaxTries(defaultRetryMaxTries))
+		return err
+	}
+	err = a.queryClusterWithRetry(ctx, query)
 	if err != nil {
 		return "", err
 	}
@@ -183,18 +165,11 @@ func (a *saramaAdminClient) GetAllTopicsMeta(ctx context.Context) (map[string]To
 		err    error
 	)
 
-	err = retry.Do(ctx, func() error {
-		a.mu.Lock()
-		defer a.mu.Unlock()
+	query := func() error {
 		topics, err = a.client.ListTopics()
-		if err == nil {
-			return nil
-		}
-		if !errors.Is(err, syscall.EPIPE) {
-			return err
-		}
-		return a.reset()
-	}, retry.WithBackoffBaseDelay(defaultRetryBackoff), retry.WithMaxTries(defaultRetryMaxTries))
+		return err
+	}
+	err = a.queryClusterWithRetry(ctx, query)
 	if err != nil {
 		return nil, err
 	}
@@ -227,20 +202,12 @@ func (a *saramaAdminClient) GetTopicsMeta(
 		metaList []*sarama.TopicMetadata
 		err      error
 	)
-
-	err = retry.Do(ctx, func() error {
-		a.mu.Lock()
-		defer a.mu.Unlock()
+	query := func() error {
 		metaList, err = a.client.DescribeTopics(topics)
-		if err == nil {
-			return nil
-		}
-		if !errors.Is(err, syscall.EPIPE) {
-			return err
-		}
+		return err
+	}
 
-		return a.reset()
-	}, retry.WithBackoffBaseDelay(defaultRetryBackoff), retry.WithMaxTries(defaultRetryMaxTries))
+	err = a.queryClusterWithRetry(ctx, query)
 	if err != nil {
 		return nil, err
 	}
@@ -274,23 +241,10 @@ func (a *saramaAdminClient) CreateTopic(
 		NumPartitions:     detail.NumPartitions,
 		ReplicationFactor: detail.ReplicationFactor,
 	}
-	err := retry.Do(ctx, func() error {
-		a.mu.Lock()
-		defer a.mu.Unlock()
-		err := a.client.CreateTopic(detail.Name, request, validateOnly)
-		if err == nil {
-			return nil
-		}
-		if strings.Contains(err.Error(), sarama.ErrTopicAlreadyExists.Error()) {
-			return nil
-		}
-		if !errors.Is(err, syscall.EPIPE) {
-			return err
-		}
-		return a.reset()
-	}, retry.WithBackoffBaseDelay(defaultRetryBackoff), retry.WithMaxTries(defaultRetryMaxTries))
-
-	return err
+	query := func() error {
+		return a.client.CreateTopic(detail.Name, request, validateOnly)
+	}
+	return a.queryClusterWithRetry(ctx, query)
 }
 
 func (a *saramaAdminClient) Close() error {

--- a/pkg/sink/kafka/admin.go
+++ b/pkg/sink/kafka/admin.go
@@ -71,15 +71,11 @@ func (a *saramaAdminClient) reset() error {
 	return errors.New("retry after reset")
 }
 
-func (a *saramaAdminClient) GetAllBrokers(ctx context.Context) ([]Broker, error) {
-	var (
-		brokers []*sarama.Broker
-		err     error
-	)
-	err = retry.Do(ctx, func() error {
+func (a *saramaAdminClient) queryClusterWithRetry(ctx context.Context, query func() error) error {
+	err := retry.Do(ctx, func() error {
 		a.mu.Lock()
 		defer a.mu.Unlock()
-		brokers, _, err = a.client.DescribeCluster()
+		err := query()
 		if err == nil {
 			return nil
 		}
@@ -90,6 +86,20 @@ func (a *saramaAdminClient) GetAllBrokers(ctx context.Context) ([]Broker, error)
 
 		return a.reset()
 	}, retry.WithBackoffBaseDelay(defaultRetryBackoff), retry.WithMaxTries(defaultRetryMaxTries))
+	return err
+}
+
+func (a *saramaAdminClient) GetAllBrokers(ctx context.Context) ([]Broker, error) {
+	var (
+		brokers []*sarama.Broker
+		err     error
+	)
+	query := func() error {
+		brokers, _, err = a.client.DescribeCluster()
+		return err
+	}
+
+	err = a.queryClusterWithRetry(ctx, query)
 	if err != nil {
 		return nil, err
 	}

--- a/pkg/sink/kafka/admin.go
+++ b/pkg/sink/kafka/admin.go
@@ -36,6 +36,11 @@ type saramaAdminClient struct {
 	client sarama.ClusterAdmin
 }
 
+const (
+	defaultRetryBackoff  = 20
+	defaultRetryMaxTries = 3
+)
+
 // NewSaramaAdminClient constructs a ClusterAdminClient with sarama.
 func NewSaramaAdminClient(ctx context.Context, config *Options) (ClusterAdminClient, error) {
 	saramaConfig, err := NewSaramaConfig(ctx, config)
@@ -63,7 +68,7 @@ func (a *saramaAdminClient) reset() error {
 	_ = a.client.Close()
 	a.client = newClient
 
-	return nil
+	return errors.New("retry after reset")
 }
 
 func (a *saramaAdminClient) GetAllBrokers(ctx context.Context) ([]Broker, error) {
@@ -84,7 +89,7 @@ func (a *saramaAdminClient) GetAllBrokers(ctx context.Context) ([]Broker, error)
 		}
 
 		return a.reset()
-	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
+	}, retry.WithBackoffBaseDelay(defaultRetryBackoff), retry.WithMaxTries(defaultRetryMaxTries))
 	if err != nil {
 		return nil, err
 	}
@@ -117,7 +122,7 @@ func (a *saramaAdminClient) GetCoordinator(ctx context.Context) (int, error) {
 		}
 
 		return a.reset()
-	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
+	}, retry.WithBackoffBaseDelay(defaultRetryBackoff), retry.WithMaxTries(defaultRetryMaxTries))
 
 	return int(controllerID), err
 }
@@ -148,7 +153,7 @@ func (a *saramaAdminClient) GetBrokerConfig(
 			return err
 		}
 		return a.reset()
-	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
+	}, retry.WithBackoffBaseDelay(defaultRetryBackoff), retry.WithMaxTries(defaultRetryMaxTries))
 	if err != nil {
 		return "", err
 	}
@@ -179,7 +184,10 @@ func (a *saramaAdminClient) GetAllTopicsMeta(ctx context.Context) (map[string]To
 			return err
 		}
 		return a.reset()
-	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
+	}, retry.WithBackoffBaseDelay(defaultRetryBackoff), retry.WithMaxTries(defaultRetryMaxTries))
+	if err != nil {
+		return nil, err
+	}
 
 	result := make(map[string]TopicDetail, len(topics))
 	for topic, detail := range topics {
@@ -222,7 +230,7 @@ func (a *saramaAdminClient) GetTopicsMeta(
 		}
 
 		return a.reset()
-	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
+	}, retry.WithBackoffBaseDelay(defaultRetryBackoff), retry.WithMaxTries(defaultRetryMaxTries))
 	if err != nil {
 		return nil, err
 	}
@@ -270,7 +278,7 @@ func (a *saramaAdminClient) CreateTopic(
 			return err
 		}
 		return a.reset()
-	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
+	}, retry.WithBackoffBaseDelay(defaultRetryBackoff), retry.WithMaxTries(defaultRetryMaxTries))
 
 	return err
 }

--- a/pkg/sink/kafka/admin.go
+++ b/pkg/sink/kafka/admin.go
@@ -15,8 +15,10 @@ package kafka
 
 import (
 	"context"
+	"errors"
 	"strconv"
 	"strings"
+	"syscall"
 
 	"github.com/Shopify/sarama"
 	"github.com/pingcap/log"
@@ -25,7 +27,9 @@ import (
 )
 
 type saramaAdminClient struct {
-	client sarama.ClusterAdmin
+	brokerEndpoints []string
+	config          *sarama.Config
+	client          sarama.ClusterAdmin
 }
 
 // NewSaramaAdminClient constructs a ClusterAdminClient with sarama.
@@ -39,7 +43,11 @@ func NewSaramaAdminClient(ctx context.Context, config *Options) (ClusterAdminCli
 	if err != nil {
 		return nil, err
 	}
-	return &saramaAdminClient{client: client}, nil
+	return &saramaAdminClient{
+		client:          client,
+		brokerEndpoints: config.BrokerEndpoints,
+		config:          saramaConfig,
+	}, nil
 }
 
 func (a *saramaAdminClient) GetAllBrokers(context.Context) ([]Broker, error) {
@@ -90,12 +98,30 @@ func (a *saramaAdminClient) GetBrokerConfig(_ context.Context, configName string
 	return configEntries[0].Value, nil
 }
 
-func (a *saramaAdminClient) GetAllTopicsMeta(context.Context) (map[string]TopicDetail, error) {
+func (a *saramaAdminClient) getAllTopicsMeta() (map[string]sarama.TopicDetail, error) {
 	topics, err := a.client.ListTopics()
+	if err == nil {
+		return topics, nil
+	}
+
+	if !errors.Is(err, syscall.EPIPE) {
+		return nil, err
+	}
+
+	client, err := sarama.NewClusterAdmin(a.brokerEndpoints, a.config)
 	if err != nil {
 		return nil, err
 	}
 
+	a.client = client
+	return a.client.ListTopics()
+}
+
+func (a *saramaAdminClient) GetAllTopicsMeta(context.Context) (map[string]TopicDetail, error) {
+	topics, err := a.getAllTopicsMeta()
+	if err != nil {
+		return nil, cerror.Trace(err)
+	}
 	result := make(map[string]TopicDetail, len(topics))
 	for topic, detail := range topics {
 		configEntries := make(map[string]string, len(detail.ConfigEntries))

--- a/pkg/sink/kafka/admin.go
+++ b/pkg/sink/kafka/admin.go
@@ -18,18 +18,22 @@ import (
 	"errors"
 	"strconv"
 	"strings"
+	"sync"
 	"syscall"
 
 	"github.com/Shopify/sarama"
 	"github.com/pingcap/log"
 	cerror "github.com/pingcap/tiflow/pkg/errors"
+	"github.com/pingcap/tiflow/pkg/retry"
 	"go.uber.org/zap"
 )
 
 type saramaAdminClient struct {
 	brokerEndpoints []string
 	config          *sarama.Config
-	client          sarama.ClusterAdmin
+
+	mu     sync.Mutex
+	client sarama.ClusterAdmin
 }
 
 // NewSaramaAdminClient constructs a ClusterAdminClient with sarama.
@@ -50,8 +54,37 @@ func NewSaramaAdminClient(ctx context.Context, config *Options) (ClusterAdminCli
 	}, nil
 }
 
-func (a *saramaAdminClient) GetAllBrokers(context.Context) ([]Broker, error) {
-	brokers, _, err := a.client.DescribeCluster()
+func (a *saramaAdminClient) reset() error {
+	newClient, err := sarama.NewClusterAdmin(a.brokerEndpoints, a.config)
+	if err != nil {
+		return cerror.Trace(err)
+	}
+
+	_ = a.client.Close()
+	a.client = newClient
+
+	return nil
+}
+
+func (a *saramaAdminClient) GetAllBrokers(ctx context.Context) ([]Broker, error) {
+	var (
+		brokers []*sarama.Broker
+		err     error
+	)
+	err = retry.Do(ctx, func() error {
+		a.mu.Lock()
+		defer a.mu.Unlock()
+		brokers, _, err = a.client.DescribeCluster()
+		if err == nil {
+			return nil
+		}
+
+		if !errors.Is(err, syscall.EPIPE) {
+			return err
+		}
+
+		return a.reset()
+	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
 	if err != nil {
 		return nil, err
 	}
@@ -66,25 +99,56 @@ func (a *saramaAdminClient) GetAllBrokers(context.Context) ([]Broker, error) {
 	return result, nil
 }
 
-func (a *saramaAdminClient) GetCoordinator(context.Context) (int, error) {
-	_, controllerID, err := a.client.DescribeCluster()
-	if err != nil {
-		return 0, err
-	}
-	return int(controllerID), nil
+func (a *saramaAdminClient) GetCoordinator(ctx context.Context) (int, error) {
+	var (
+		controllerID int32
+		err          error
+	)
+	err = retry.Do(ctx, func() error {
+		a.mu.Lock()
+		defer a.mu.Unlock()
+		_, controllerID, err = a.client.DescribeCluster()
+		if err == nil {
+			return nil
+		}
+
+		if !errors.Is(err, syscall.EPIPE) {
+			return err
+		}
+
+		return a.reset()
+	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
+
+	return int(controllerID), err
 }
 
-func (a *saramaAdminClient) GetBrokerConfig(_ context.Context, configName string) (string, error) {
-	_, controller, err := a.client.DescribeCluster()
+func (a *saramaAdminClient) GetBrokerConfig(
+	ctx context.Context,
+	configName string,
+) (string, error) {
+	controller, err := a.GetCoordinator(ctx)
 	if err != nil {
 		return "", err
 	}
 
-	configEntries, err := a.client.DescribeConfig(sarama.ConfigResource{
-		Type:        sarama.BrokerResource,
-		Name:        strconv.Itoa(int(controller)),
-		ConfigNames: []string{configName},
-	})
+	var configEntries []sarama.ConfigEntry
+	err = retry.Do(ctx, func() error {
+		a.mu.Lock()
+		defer a.mu.Unlock()
+		configEntries, err = a.client.DescribeConfig(sarama.ConfigResource{
+			Type:        sarama.BrokerResource,
+			Name:        strconv.Itoa(controller),
+			ConfigNames: []string{configName},
+		})
+		if err == nil {
+			return nil
+		}
+
+		if !errors.Is(err, syscall.EPIPE) {
+			return err
+		}
+		return a.reset()
+	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
 	if err != nil {
 		return "", err
 	}
@@ -98,30 +162,25 @@ func (a *saramaAdminClient) GetBrokerConfig(_ context.Context, configName string
 	return configEntries[0].Value, nil
 }
 
-func (a *saramaAdminClient) getAllTopicsMeta() (map[string]sarama.TopicDetail, error) {
-	topics, err := a.client.ListTopics()
-	if err == nil {
-		return topics, nil
-	}
+func (a *saramaAdminClient) GetAllTopicsMeta(ctx context.Context) (map[string]TopicDetail, error) {
+	var (
+		topics map[string]sarama.TopicDetail
+		err    error
+	)
 
-	if !errors.Is(err, syscall.EPIPE) {
-		return nil, err
-	}
+	err = retry.Do(ctx, func() error {
+		a.mu.Lock()
+		defer a.mu.Unlock()
+		topics, err = a.client.ListTopics()
+		if err == nil {
+			return nil
+		}
+		if !errors.Is(err, syscall.EPIPE) {
+			return err
+		}
+		return a.reset()
+	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
 
-	client, err := sarama.NewClusterAdmin(a.brokerEndpoints, a.config)
-	if err != nil {
-		return nil, err
-	}
-
-	a.client = client
-	return a.client.ListTopics()
-}
-
-func (a *saramaAdminClient) GetAllTopicsMeta(context.Context) (map[string]TopicDetail, error) {
-	topics, err := a.getAllTopicsMeta()
-	if err != nil {
-		return nil, cerror.Trace(err)
-	}
 	result := make(map[string]TopicDetail, len(topics))
 	for topic, detail := range topics {
 		configEntries := make(map[string]string, len(detail.ConfigEntries))
@@ -142,11 +201,28 @@ func (a *saramaAdminClient) GetAllTopicsMeta(context.Context) (map[string]TopicD
 }
 
 func (a *saramaAdminClient) GetTopicsMeta(
-	_ context.Context,
+	ctx context.Context,
 	topics []string,
 	ignoreTopicError bool,
 ) (map[string]TopicDetail, error) {
-	metaList, err := a.client.DescribeTopics(topics)
+	var (
+		metaList []*sarama.TopicMetadata
+		err      error
+	)
+
+	err = retry.Do(ctx, func() error {
+		a.mu.Lock()
+		defer a.mu.Unlock()
+		metaList, err = a.client.DescribeTopics(topics)
+		if err == nil {
+			return nil
+		}
+		if !errors.Is(err, syscall.EPIPE) {
+			return err
+		}
+
+		return a.reset()
+	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
 	if err != nil {
 		return nil, err
 	}
@@ -172,21 +248,35 @@ func (a *saramaAdminClient) GetTopicsMeta(
 }
 
 func (a *saramaAdminClient) CreateTopic(
-	_ context.Context,
+	ctx context.Context,
 	detail *TopicDetail,
 	validateOnly bool,
 ) error {
-	err := a.client.CreateTopic(detail.Name, &sarama.TopicDetail{
+	request := &sarama.TopicDetail{
 		NumPartitions:     detail.NumPartitions,
 		ReplicationFactor: detail.ReplicationFactor,
-	}, validateOnly)
-	// Ignore the already exists error because it's not harmful.
-	if err != nil && !strings.Contains(err.Error(), sarama.ErrTopicAlreadyExists.Error()) {
-		return err
 	}
-	return nil
+	err := retry.Do(ctx, func() error {
+		a.mu.Lock()
+		defer a.mu.Unlock()
+		err := a.client.CreateTopic(detail.Name, request, validateOnly)
+		if err == nil {
+			return nil
+		}
+		if strings.Contains(err.Error(), sarama.ErrTopicAlreadyExists.Error()) {
+			return nil
+		}
+		if !errors.Is(err, syscall.EPIPE) {
+			return err
+		}
+		return a.reset()
+	}, retry.WithBackoffBaseDelay(20), retry.WithMaxTries(1))
+
+	return err
 }
 
 func (a *saramaAdminClient) Close() error {
+	a.mu.Lock()
+	defer a.mu.Unlock()
 	return a.client.Close()
 }

